### ISSUE TITLE
Older Rubies (2.1 to 2.3 at least) don't install on default macOS image (10.13.6)

### ISSUE THREAD

**Issue by aried3r** [travis-ci/travis-rubies#68]
I've described this issue [here](https://github.com/tonytonyjan/jaro_winkler/pull/38#issuecomment-570342072) but will post it here as well since I believe it belongs here as well. Feel free to close this issue if this is not the place. I'm not sure if it's a Travis CI issue or if I should open an issue in the rvm repo.

It seems a new OpenSSL version was released on Homebrew ([1.1](https://travis-ci.org/tonytonyjan/jaro_winkler/jobs/632006062#L83), from [1.0](https://travis-ci.org/tonytonyjan/jaro_winkler/jobs/614396866#L83) (this also contains a successful and a failed test run)) and also the macOS build image on Travis [changed](https://travis-ci.org/tonytonyjan/jaro_winkler/jobs/631999174#L17) since the [last test run of `jaro_winkler`](https://travis-ci.org/tonytonyjan/jaro_winkler/jobs/614396866#L17) (check also the available Rubies etc). 
That seems to be the cause for the failures.
Somebody [posted a possible workaround](https://github.com/Homebrew/homebrew-core/issues/47368#issuecomment-564514496) to get an older OpenSSL version installed via homebrew.

Further reading:
https://travis-ci.community/t/rvm-install-fails-on-osx-due-to-missing-libraries/5596
https://github.com/rvm/rvm/issues/4819
https://github.com/rvm/rvm/issues/4781
https://github.com/rvm/rvm/issues/4463
https://github.com/Homebrew/homebrew-core/issues/46454

**Comment by BanzaiMan** [comment]
I am not sure if we can reliably support Rubies <= 2.3.x on macOS at this point.

The problem is quite complicated:

1. Rubies before 2.4.0 do not compile with OpenSSL 1.1, thus we *_need_* OpenSSL 1.0.x, which is no longer supported.
1. Homebrew, our package manager on macOS, has dropped support for OpenSSL 1.0.x, justifiably.
1. This means that we have to rely on a third-party tool (or create our own) to provide OpenSSL 1.0.x. It is possible to compile OpenSSL 1.0.x on the fly, but it will take quite some time to do so. I have not found any binary package to sidestep this issue (let alone a trustworthy one that we can use long term).
1. RVM currently is not aware of this issue, so it appears to prefer the latest OpenSSL version when we ask for Ruby < 2.4.0.
1. The problem would be mitigated if we could compile a static macOS binary that can be ported across machines, but as far as I know, currently there is no way to do so.

It is conceivably possible to:

1. Compile OpenSSL 1.0 (this is not difficult)
1. Coerce RVM to compile Ruby with this OpenSSL instead of the latest version 
1. Package all the binary files into the archive, so that users don't have to compile things at build time.
1. ???
1. $$ Profit $$

**Comment by hurricup** [comment]
As I mentioned in my mail to support, there are at least two ways to do so:

```
# using rvm pkg install (deprecated)
rvm pkg install openssl
rvm install 2.3.0 --with-openssl-dir=$HOME/.rvm/usr
```

or

```
# alternative using rbenv recipe
brew install rbenv/tap/openssl@1.0
rvm install 2.3.0 --with-openssl-dir=$(brew --prefix openssl@1.0)
```

**Comment by BanzaiMan** [comment]
@hurricup Do you have a URL to the working build?

**Comment by hurricup** [comment]
No. This is how I set up locally. But travis for some reason can't do the same. And this is really strange.

**Comment by BanzaiMan** [comment]
With
```yaml
language: generic
os: osx
osx_image: xcode10.3
before_install:
  - env HOMEBREW_NO_AUTO_UPDATE=1 brew install rbenv/tap/openssl@1.0
  - rvm reinstall 2.3.0 --with-openssl-dir=$(brew --prefix openssl@1.0)
install:
  - true
script:
  - "ruby -v -ropenssl -e \"puts OpenSSL::VERSION\""
```

You can compile OpenSSL 1.0 and Ruby 2.3.0 

This takes about 30 minutes.

**Comment by hurricup** [comment]
I presume travis could pre-build them and use like binary distribution. I know it's kinda tricky on macos, but still possible.

**Comment by BanzaiMan** [comment]
In theory, yes, but I have not found a way to play nicely with Homebrew.

**Comment by native-api** [comment]
> I have not found a way to play nicely with Homebrew.

`brew install rbenv/tap/openssl@1.0` <- doesn't this count as "nice"?

**Comment by BanzaiMan** [comment]
As in, bundling up what Homebrew writes on the build machine in such a way that can be reused elsewhere. Some of it might be simulated by caching, but it seems to fail in a very hard-to-understand ways.

Ruby 2.4 support is already dropped. I question the benefit of continuing to support 2.3, to be honest.

**Comment by hurricup** [comment]
The question is not about lifetime from devs, but usage statistics. Like https://stats.rubygems.org/

**Comment by native-api** [comment]
> As in, bundling up what Homebrew writes on the build machine in such a way that can be reused elsewhere.

Since this is RVM's problem, you either supplement its logic in travis-build (`brew install` before `rvm install`), tap into its logic (I don't know how you tap into it to use https://rubies.travis-ci.org/ since [the relevant logic is not on Github](https://github.com/search?q=org%3Atravis-ci+%22rubies.travis-ci.org%22&type=code) so can't say) -- maybe something similar, or make a PR against it to fix the problem for everyone.

Alternatively, you can use `rbenv` for affected versions who have already fixed that problem for themselves (I don't know how compatible their Ruby installation is though).
